src/org/traccar/protocol/EelinkProtocolDecoder.java

```diff
@@ -25,43 +25,60 @@
 import org.traccar.model.CellTower;
 import org.traccar.model.Network;
 import org.traccar.model.Position;
+import org.traccar.helper.Log;
 
 import java.net.SocketAddress;
 import java.util.Date;
 
+import java.nio.charset.StandardCharsets;
+
 public class EelinkProtocolDecoder extends BaseProtocolDecoder {
 
     public EelinkProtocolDecoder(EelinkProtocol protocol) {
         super(protocol);
     }
 
+    private static final short HEADER_KEY = 0x6767;
+
     public static final int MSG_LOGIN = 0x01;
     public static final int MSG_GPS = 0x02;
     public static final int MSG_HEARTBEAT = 0x03;
     public static final int MSG_ALARM = 0x04;
     public static final int MSG_STATE = 0x05;
     public static final int MSG_SMS = 0x06;
     public static final int MSG_OBD = 0x07;
+    public static final int MSG_OBD_FAULT = 0x09;
     public static final int MSG_DOWNLINK = 0x80;
     public static final int MSG_DATA = 0x81;
 
     public static final int MSG_NORMAL = 0x12;
     public static final int MSG_WARNING = 0x14;
     public static final int MSG_REPORT = 0x15;
     public static final int MSG_COMMAND = 0x16;
+
     public static final int MSG_OBD_DATA = 0x17;
     public static final int MSG_OBD_BODY = 0x18;
     public static final int MSG_OBD_CODE = 0x19;
-    public static final int MSG_CAMERA_INFO = 0x1E;
-    public static final int MSG_CAMERA_DATA = 0x1F;
 
-    private void sendResponse(Channel channel, int type, int index) {
+    public static final int MSG_CAMERA_INFO = 0x0E;
+    public static final int MSG_CAMERA_DATA = 0x0F;
+
+    public static final int MSG_SIGN_COMMAND = 0x01;
+    public static final int MSG_SIGN_NEWS = 0x02;
+
+    private void sendResponse(Channel channel, int type, int index, ChannelBuffer buf) {
         if (channel != null) {
-            ChannelBuffer response = ChannelBuffers.buffer(7);
-            response.writeByte(0x67); response.writeByte(0x67); // header
+            ChannelBuffer response = ChannelBuffers.dynamicBuffer();
+            response.writeShort(HEADER_KEY);
             response.writeByte(type);
-            response.writeShort(2); // length
+            response.writeShort(2); // total packet length
             response.writeShort(index);
+
+            if (buf != null && buf.readableBytes() > 0) {
+                response.writeBytes(buf);
+                response.setShort(3, 2 + buf.writerIndex()); // change packet length including buf
```

Comment:
I could but but would need to do the same when writing out the content too which can be null on ACK packets;
It would still end up with a if statement when writing the content so you dont try and pass null to writeBytes.